### PR TITLE
ENH add gap safe rules

### PR DESCRIPTION
**Pull request by PABannier**
The goal of this PR is to add the `GSROptim` solver from https://github.com/EugeneNdiaye/Gap_Safe_Rules

### CODE REVIEW

File: solvers/gsroptim.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        self.lambdas = lambdas
+        self.fit_intercept = fit_intercept
+
+    def run(self, n_iter):
+        warnings.filterwarnings("ignore", category=ConvergenceWarning)
+        self.coefs = lasso_path(self.X, self.y, self.lambdas, eps=1e-12,
+                                max_iter=n_iter, fit_intercept=self.fit_intercept)[1]
+
+    @staticmethod
+    def get_next(previous):
```
Comment (mathurinm) [inline solvers/gsroptim.py]: I don't think this growth should be used, in GSR this should correspond to CD epochs hence we want a faster growth

File: solvers/gsroptim.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    requirements = [
+        'pip:git+https://github.com/EugeneNdiaye/Gap_Safe_Rules.git@master'
+    ]
+    references = [
+        "E. Ndiaye, O. Fercoq, A. Gramfort and J. Salmon, JMLR, "
+        '"Gap Safe screening rules for sparsity enforcing penalties", '
+        "vol. 18, pp. 1-33 (2017)"
+    ]
+
+    def skip(self, X, y, lambdas, fit_intercept):
```
Comment (mathurinm) [inline solvers/gsroptim.py]: are you sure that sparse data is not supported ?

Comment (PABannier) [reply solvers/gsroptim.py]: i've checked the docstring. It expects an array of shape `(n_samples, n_features)`. 
I don't think it is optimized for sparse matrices.

Comment (mathurinm) [reply solvers/gsroptim.py]: You are right but the docstring is wrong. sparse X is supported 
https://github.com/EugeneNdiaye/Gap_Safe_Rules/blob/master/gsroptim/cd_lasso_fast.pyx#L156 
https://github.com/EugeneNdiaye/Gap_Safe_Rules/blob/master/gsroptim/lasso.py#L110

Comment (PABannier) [reply solvers/gsroptim.py]: Ok, good news then. I've removed the `skip` method.

**Review by mathurinm — APPROVED** [review]
Tested on leukemia, runs fine

**Comment by jolars** [comment]
Looks good to me here too.

image

**Comment by jolars** [comment]
Thanks a lot for the work here @PABannier !

**Pull request merged**